### DISCUSSION TITLE
AST rendering for Handlebars

### DISCUSSION THREAD

**Discussion by HuJingKang** [handlebars-lang/handlebars.js#1869] [Ideas]
Hi, I am using Handlebars for SSR in NodeJS.
By reading the source code, I learned that the rendering process is:
AST -> opcodes -> templateSpec -> eval(templateSpec) -> result.
Since the eval operation is time-consuming, I guess that if the rendering process is optimized to AST -> result, the performance may be greatly improved.
Could you please evaluate whether this optimization is feasible? Or is this an implementation that you have considered but was abandoned for some reason?

**Comment by jaylinski** [comment]
This may be possible, but it would probably be a very time-consuming and complex task.

**Comment by HuJingKang** [comment]
@jaylinski Recently I wrote a simple demo for this. Could you please help me to review the code and make some suggestions?
https://github.com/HuJingKang/fast-handlebars

**Reply by jaylinski** [reply]
I can take a look at it. What are the things you need input on?

**Reply by HuJingKang** [reply]
Thank you so much. Usage and input are no different from Handlebars. You can look at this: 
https://github.com/HuJingKang/fast-handlebars/blob/main/examples/index.ts

**Reply by rickyes** [reply]
Hi, @jaylinski. We have refactored parser/render based on AST and benchmarked it to improve performance by about `3x`, so you can help see if it can be incorporated into `handlebars.js` repository as an experimental feature.
